sdk/samples/iot/sample_pnp.c

```diff
@@ -0,0 +1,428 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "sample_pnp.h"
+
+#include <stdint.h>
+#include <stdio.h>
+
+#include <azure/core/az_json.h>
+#include <azure/core/az_result.h>
+#include <azure/core/az_span.h>
+#include <azure/iot/az_iot_hub_client.h>
+
+#define JSON_DOUBLE_DIGITS 2
+
+// Property values
+static char pnp_properties_buffer[64];
+
+// Device twin values
+static const az_span component_telemetry_prop_span = AZ_SPAN_LITERAL_FROM_STR("$.sub");
+static const az_span desired_temp_response_value_name = AZ_SPAN_LITERAL_FROM_STR("value");
+static const az_span desired_temp_ack_code_name = AZ_SPAN_LITERAL_FROM_STR("ac");
+static const az_span desired_temp_ack_version_name = AZ_SPAN_LITERAL_FROM_STR("av");
+static const az_span desired_temp_ack_description_name = AZ_SPAN_LITERAL_FROM_STR("ad");
+static const az_span component_specifier_name = AZ_SPAN_LITERAL_FROM_STR("__t");
+static const az_span component_specifier_value = AZ_SPAN_LITERAL_FROM_STR("c");
+static const az_span command_separator = AZ_SPAN_LITERAL_FROM_STR("*");
+static const az_span sample_iot_hub_twin_desired_version = AZ_SPAN_LITERAL_FROM_STR("$version");
+static const az_span sample_iot_hub_twin_desired = AZ_SPAN_LITERAL_FROM_STR("desired");
+
+// Visit each valid property for the component
+static az_result visit_component_properties(
+    az_span component_name,
+    az_json_reader* json_reader,
+    int32_t version,
+    char* scratch_buf,
+    int32_t scratch_buf_len,
+    pnp_property_callback property_callback,
+    void* context_ptr)
+{
+  int32_t len;
+
+  while (az_succeeded(az_json_reader_next_token(json_reader)))
+  {
+    if (json_reader->token.kind == AZ_JSON_TOKEN_PROPERTY_NAME)
+    {
+      if (az_json_token_is_text_equal(&(json_reader->token), component_specifier_name)
+          || az_json_token_is_text_equal(
+              &(json_reader->token), sample_iot_hub_twin_desired_version))
+      {
+        if (az_failed(az_json_reader_next_token(json_reader)))
+        {
+          printf("Failed to next token\r\n");
+          return AZ_ERROR_UNEXPECTED_CHAR;
+        }
+        continue;
+      }
+
+      if (az_failed(az_json_token_get_string(
+              &(json_reader->token), (char*)scratch_buf, (int32_t)scratch_buf_len, (int32_t*)&len)))
+      {
+        printf("Failed to get string property value\r\n");
+        return AZ_ERROR_UNEXPECTED_CHAR;
+      }
+
+      if (az_failed(az_json_reader_next_token(json_reader)))
+      {
+        printf("Failed to get next token\r\n");
+        return AZ_ERROR_UNEXPECTED_CHAR;
+      }
+
+      property_callback(
+          component_name,
+          az_span_init((uint8_t*)scratch_buf, len),
+          &(json_reader->token),
+          version,
+          context_ptr);
```

Comment:
I am having trouble understanding the flow here. Do we need to give the unescaped JSON string token as a span back to the user, or is giving them just the unmodified slice of the input they provided, good enough?

If it's the latter, just save and return back the token OR the span from the token, `&(json_reader->token.slice`.

Reply:
Follow up snippet from offline conversation.

Returning `az_span property_name` in the callback to the user (currently in PR):
1) az_span_is_content_equal(property_name, AZ_SPAN_FROM_STR("MY LITERAL"));
2) Not needed, because you, the pnp sample, already did it for me.

Returning `az_json_token property_name` in the callback to the user (proposal):
1) az_json_token_text_equal(property_name, AZ_SPAN_FROM_STR("MY LITERAL"))
2) az_json_token_get_string(..., scratch buffer)

Reply:
how will this will look if property name is in discontinuous memory? 

---
In reply to: [465991913](https://github.com/Azure/azure-sdk-for-c/pull/978#discussion_r465991913) [](ancestors = 465991913)

Reply:
I guess your token API will take care of it.. right?

---
In reply to: [466077550](https://github.com/Azure/azure-sdk-for-c/pull/978#discussion_r466077550) [](ancestors = 466077550,465991913)

Reply:
Yes, but atm, this code is only working on a contiguous span buffer, and it owns the creation of the reader, so that isn't a concern here.